jwst/extract_2d/grisms.py

```diff
@@ -4,20 +4,213 @@
+                       extract_orders=None):
+    """
+    Extract the spectrum for a NIRCAM TSO observation.
+
+    Parameters
+    ----------
+    input_model : jwst.datamodels.CubeModel
+
+    reference_files : dict
+        Needs to include the name of the wavelengthrange reference file
+
+    extract_height: int
+        The extraction height, in total, for the spectra in the
+        cross-dispersion direction. If this is other than None,
+        it will override the team default of 64 pixels. The team
+        wants the source centered near row 34, so the extraction
+        height is not the same on either size of the central pixel.
+
+    extract_orders: list[ints]
+        This is an optional parameter that will override the
+        orders specified for extraction in the wavelengthrange
+        reference file.
+
+    Returns
+    -------
+    output_model : jwst.datamodels.MultiSlitModel()
+
+
+    Notes
+    -----
+    This method supports NRC_TSGRISM only, where only one
+    bright object is considered in the field, so there's
+    no catalog of sources and the object is assumed to
+    have been observed at the aperture location crpix1/2.
+
+    GrismObject is a named tuple which contains distilled
+    information about each catalog object and the bounding
+    boxes that will be used to define the 2d extraction area.
+
+    Since this mode has a single known source location the utilities
+    used in the WFSS modes are overkill, instead, similar structures
+    are created during the extrac 2d process and then directly used.
+
+    https://jwst-docs.stsci.edu/display/JTI/NIRCam+Grism+Time+Series
+    """
+
+    if not isinstance(reference_files, dict):
+        raise TypeError("Expected a dictionary for reference_files")
+    if 'wavelengthrange' not in reference_files.keys():
+        raise KeyError("No wavelengthrange reference file specified")
+
+    if not isinstance(input_model, CubeModel):
+        raise TypeError('The input data model is not a CubeModel.')
+
+    if extract_height is None:
+        extract_height = 64  # set by the teams
+
+    # Get the disperser parameters which have the wave limits
+    with WavelengthrangeModel(reference_files['wavelengthrange']) as f:
+        if (f.meta.instrument.name != 'NIRCAM'):
+            raise ValueError("Wavelengthrange reference file not for NIRCAM!")
+        wavelengthrange = f.wavelengthrange
+        ref_extract_orders = f.extract_orders
+
+    # this supports the user override
+    if extract_orders is None:
+        extract_orders = ref_extract_orders
+
+    available_orders = [x[1] for x in extract_orders if x[0] == input_model.meta.instrument.filter].pop()
+    if len(available_orders) > 1:
+        log.info("Extracting into a MultiSlitModel")
+        output_model = datamodels.MultiSlitModel()
+        output_model.update(input_model)
+    else:
+        log.info("Extracting into a SlitModel")
+        output_model = datamodels.SlitModel()
```

Comment:
Is this the way it's done for NRS_BRIGHTOBJ mode too - the output is a SlitModel, as opposed to a MultiSlitModel that contains just 1 Slit instance? I just want to make sure we're consistent, so that extract_1d is able to handle both NRS_BRIGHTOBJ and NRC_TSGRISM the same way.

Reply:
`extract_1d` tests for CubeModel (the old model for NRS_BRIGHTOBJ) and SlitModel.  So NRC_TSGRISM should be handled by the same section, yes.

Reply:
that was my understanding. Hold off on reviews because after todays discussion, almost all of this code is changing. The only output will be a SlitModel with a single data cube.